node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeRepository.java

```diff
@@ -468,28 +468,37 @@ private Node move(Node node, Node.State toState, Agent agent, Optional<String> r
     }
 
     /**
-     * Removes a node. A node must be in a legal state before it can be removed.
+     * Removes all the nodes that are children of hostname before finally removing the hostname itself.
+     *
+     * @return List of all the nodes that have been removed
      */
-    public void remove(String hostname) {
-        Node nodeToRemove = getNode(hostname).orElseThrow(() ->  new NotFoundException("No node with hostname \"" + hostname + '"'));
-        List<Node.State> legalStates = dynamicAllocationEnabled() ?
-                Arrays.asList(Node.State.provisioned, Node.State.failed, Node.State.parked, Node.State.dirty) :
-                Arrays.asList(Node.State.provisioned, Node.State.failed, Node.State.parked);
+    public List<Node> removeRecursively(String hostname) {
+        Node node = getNode(hostname).orElseThrow(() -> new NotFoundException("No node with hostname \"" + hostname + '"'));
 
-        if (! legalStates.contains(nodeToRemove.state())) {
-            throw new IllegalArgumentException("Can only remove node from following states: " +
-                    legalStates.stream().map(Node.State::name).collect(Collectors.joining(", ")));
+        try (Mutex lock = lockUnallocated()) {
+            List<Node> removed = getChildNodes(hostname).stream()
+                    .filter(this::allowedToRemove)
+                    .collect(Collectors.toList());
+            if (allowedToRemove(node)) removed.add(node);
+            db.removeNodes(removed);
+
+            return removed;
+        } catch (RuntimeException e) {
+            throw new IllegalArgumentException("Failed to delete " + node.hostname(), e);
         }
+    }
 
-        if (nodeToRemove.state().equals(Node.State.dirty)) {
-            if (!(nodeToRemove.flavor().getType().equals(Flavor.Type.DOCKER_CONTAINER))) {
-                throw new IllegalArgumentException("Only docker nodes can be deleted from state dirty");
-            }
-        }
+    private boolean allowedToRemove(Node nodeToRemove) {
+        List<Node.State> legalStates = nodeToRemove.flavor().getType() == Flavor.Type.DOCKER_CONTAINER ?
+                Arrays.asList(Node.State.provisioned, Node.State.ready) :
```

Comment:
I don't think this makes sense - we should be able to delete a Docker host (including its children) even if some of the docker nodes are in state failed or parked. With hardware failure, there's little chance of the docker nodes being ready, and that wouldn't make sense either.

But, we should be able to delete any single docker node iff it is ready (I don't think we should batch provisioned in the same category).

Reply:
Fixed.